### ISSUE TITLE
Clean up taxon name lexicons

### ISSUE THREAD

**Issue by kueda** [inaturalist/inaturalist#2675]
They're a mess, b/c we allow people to enter anything.

- [ ] Add a `unique_lexicon` column to the `taxon_names` table, give it a unique index, populate it with `parameterize(lexicon)` on save, and add a `validates_uniqueness_of` validation for it. When a user tries to add a lexicon that doesn't pass this validation, show them the lexicon they were about the duplicate
- [ ] Add a validation to check if `parameterize(lexicon)` is blank, and if so, remind the user that lexicons are supposed to be in English and translations should be handled like other translated text, i.e. "Japanese" instead of "日本語"
- [ ] Reject any new names with lexicons that match any translations of the `lexicons` translation keys, e.g. if a user tries to enter "Anglais" as a lexicon, it should fail b/c `I18n.t( "lexicons.english", locale: "fr" ) == "Anglais" && I18n.t( "lexicons.english", locale: "en" ) != "Anglais"`
- [ ] Clean up / combine all the lexicons that don't pass these new rules (this is the hard part)

**Comment by todtb** [comment]
I'd gladly to take a look at this, but I'm a bit confused by the desired functionality.

We could add (with null for the data migration stage):
```ruby
add_column :taxon_names, :unique_lexicon, :string
add_index :taxon_names, :unique_lexicon, unique: true, where: "unique_lexicon IS NOT NULL"
```
but this means we would only support a single `taxon_name` per lexicon. 

Should this be a composite index of lexicon and name instead? It couldn't be on taxon_id as we still want to support multiple common names in each lexicon right?

Also if adding a new lexicon such as "Ido", this would be allowed since there is no lexicon translation key for the language? How do we differentiate this from "伊多語"?

Ultimately the goal is to avoid something like this?
| id  | name  | lexicon  | taxon_id  | 
|---|---|---|---|
| 1  |  Fungi |  German | 9  |
| 2  | Pilze  |  German  |  9  | 
| 3 | Pilze  |  Deutsch  |  9 |
| 4 | Pilze  |  deutsche sprache  |  9 |

**Comment by kueda** [comment]
> Should this be a composite index of lexicon and name instead? It couldn't be on taxon_id as we still want to support multiple common names in each lexicon right?

Yeah, that sounds better. Sorry, I clearly didn't think that through very well.

> Also if adding a new lexicon such as "Ido", this would be allowed since there is no lexicon translation key for the language? How do we differentiate this from "伊多語"?

I don't think we can, unless you have any good ideas. We would need to have some way of storing Romanized versions of lexicons in other character sets, like a `zh-CN-roman` locale or something. It's not clear to me that that's a huge problem yet, though.

> Ultimately the goal is to avoid something like this?

Yeah, or to use a real example,

```
                      lexicon                      |  count  
---------------------------------------------------+---------
 Chinese (Traditional)                             |   23853
 Chinese Traditional                               |     414
```

**Comment by kueda** [comment]
> After validating english_lexicon_if_exists, perform a validation under similar conditions that checks the parameterized version of the submitted value is not in the distinct set of parameterized lexicons already persisted

This sounds like the way to me. A new table sounds like overkill to me.

And yeah, lots of weird and duplicate names from various imports over the years, though generally not weird lexicons from the imports on our end, at least. Most name importing we do is for a single lexicon, so we tend to use the conventional lexicon and not "Español (Argentina)"

**Comment by todtb** [comment]
Just want to follow up on this @kueda. I think the major tasks here are complete, with the exception of "Clean up / combine all the lexicons that don't pass these new rules (this is the hard part)".

I'm not sure how I can help with that bit, but if there's something I can do, let me know and I'd be glad to contribute where possible.
